src/main/java/com/mongodb/spark/sql/connector/schema/InferSchema.java

```diff
@@ -0,0 +1,348 @@
+import org.apache.spark.sql.types.DecimalType;
+import org.apache.spark.sql.types.DoubleType;
+import org.apache.spark.sql.types.IntegerType;
+import org.apache.spark.sql.types.LongType;
+import org.apache.spark.sql.types.MapType;
+import org.apache.spark.sql.types.Metadata;
+import org.apache.spark.sql.types.StructField;
+import org.apache.spark.sql.types.StructType;
+import org.apache.spark.sql.util.CaseInsensitiveStringMap;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+import org.jetbrains.annotations.VisibleForTesting;
+
+import org.bson.BsonDocument;
+import org.bson.BsonValue;
+
+import com.mongodb.client.model.Aggregates;
+
+import com.mongodb.spark.sql.connector.assertions.Assertions;
+import com.mongodb.spark.sql.connector.config.MongoConfig;
+import com.mongodb.spark.sql.connector.config.ReadConfig;
+
+/**
+ * A helper that determines the {@code StructType} for a {@code BsonDocument} and finds the common
+ * {@code StructType} for a list of BsonDocuments.
+ *
+ * <p>All Bson types are considered convertible to Spark types. For any Bson types that doesn't have
+ * a direct conversion to a Spark type then a String type will be used.
+ */
+@NotNull
+public final class InferSchema {
+
+  /**
+   * Infer the schema for the collection
+   *
+   * @param options the configuration options to determine the namespace to determine the schema for
+   * @return the schema
+   */
+  public static StructType inferSchema(final CaseInsensitiveStringMap options) {
+    ReadConfig readConfig = MongoConfig.readConfig(options.asCaseSensitiveMap());
+    return inferSchema(
+        readConfig.withCollection(
+            coll ->
+                coll.aggregate(
+                        singletonList(Aggregates.sample(readConfig.getInferSchemaSampleSize())))
+                    .into(new ArrayList<>())),
+        readConfig);
+  }
+
+  @VisibleForTesting
+  static StructType inferSchema(
+      final List<BsonDocument> bsonDocuments, final ReadConfig readConfig) {
+    return bsonDocuments.stream()
+        .map(d -> getStructType(d, readConfig))
+        .reduce(PLACE_HOLDER_STRUCT_TYPE, (dt1, dt2) -> compatibleStructType(dt1, dt2, readConfig));
+  }
+
+  @NotNull
+  private static StructType getStructType(
+      final BsonDocument bsonDocument, final ReadConfig readConfig) {
+    return (StructType) getDataType(bsonDocument, readConfig);
+  }
+
+  @VisibleForTesting
+  static DataType getDataType(final BsonValue bsonValue, final ReadConfig readConfig) {
+    switch (bsonValue.getBsonType()) {
+      case DOCUMENT:
+        List<StructField> fields = new ArrayList<>();
+        bsonValue
+            .asDocument()
+            .forEach(
+                (k, v) ->
+                    fields.add(
+                        new StructField(k, getDataType(v, readConfig), true, Metadata.empty())));
+        return dataTypeCheckStructTypeToMapType(DataTypes.createStructType(fields), readConfig);
```

Comment:
Just a question: why do we represent some documents as `StructType` and some as `MapType` depending on how many fields (`INFER_SCHEMA_MAP_TYPE_MINIMUM_KEY_SIZE_CONFIG`) the document has?

Reply:
Good question - this feature was added as a user request in the original connector.  

Their data was very dynamic and had many different key names. So trying to merge them into a single StructType was expensive in the size of the Structs and error prone as the sampling wouldn't pick up all the variations in their map like data.

Reply:
So `MapType` is more versatile than `StructType`, got it.